### PR TITLE
revisions to #3197

### PR DESCRIPTION
**Pull request by mattjj**
The previous version of `core.find_top_trace` was written and tested for performance; the revised version of #3197 caused at least a 25% slowdown in the dispatch time of `lax.add(1, 2)` (and so likely a much bigger slowdown for the find_top_trace timing alone).

Moreover, the check isn't so natural in core.py. Instead, we can just change the error message in xla.abstractify, since invalid types lead to abstractification errors when we apply primitive impls.

Revises #3197

### CODE REVIEW

File: tests/api_test.py [hunk 1]
```diff
@@ -1765,6 +1765,12 @@ def f(a, b):
     with self.assertRaisesRegex(ValueError, msg):
       g(jnp.ones((1, 1)), b=1)
 
+  def test_primitive_jaxtype_error(self):
+    with self.assertRaisesRegex(
+        TypeError, "Argument .* of type .* is not a valid JAX type"):
+      lax.add(1, 'hi')
```
Comment (gnecula) [inline tests/api_test.py:1771]: I am surprised that the error in xla.abstractify is triggered here. I can't quite tell how. Probably by way of `core._valid_jaxtype` which is only called by `core._check_arg` but then I do not see how `lax.add` will call _check_arg. Is `lax.add` a compiled primitive? 

I see now, all arithmetic operators are compiled. This solution then does not work for the non-compiled lax primitives.

**Review by gnecula — APPROVED** [review]
Did you intentionally drop 
```
assert skip_checks or all(isinstance(arg, Tracer)
                              or valid_jaxtype(arg) for arg in args), args
```

from bind? Also for speed?

**Comment by mattjj** [comment]
> Did you intentionally drop [...] from bind

No, that was not intentional. This PR didn't edit bind, but I see now that #3197 did, and I failed to restore that bit. (Things gated by skip_checks can be expensive!)

Thanks for the catch!

File: tests/api_test.py (hunk 1 above)
Comment (mattjj) [reply tests/api_test.py:1771]: Right yeah. In op-by-op mode the impls should check the types of their arguments, and any impl based on `xla.apply_primitive` needs to go through xla and hence `xla.abstractify`. In jaxpr tracing, the type checking should happen in the abstract eval rules.

**Pull request merged**